### ISSUE TITLE
StaleElementReferenceError when trying to access ElementCollection item

### ISSUE THREAD

**Issue by maybelinot** [watir/watir#571]
I'm getting `StaleElementReferenceError` error when trying to access element of `ElementCollection` which was changed/reloaded after first call was made. Unfortunately I can't share example to reproduce this issue. As possible solutions I could think of: 
* need to remove all stale elements at the moment we converting collection to an array
* convert stale elements to StaleElement objects (when user trying to access item of Collection)

**Firefox version**: 50.1.0
**Watir version**: 6.4.0.rc1
```
[1] pry> res.kind_of? Watir::ElementCollection
=> true
[2] pry> res
=> #<Watir::HTMLElementCollection:0x00000001b8a8e8
 @elements=[#<Selenium::WebDriver::Element:0x..f8103bdd0046f0a86 id="{0539143b-665c-4111-918a-8f408d5ace87}">],
 @query_scope=#<Watir::Browser:0x2d8d21d20f2bbd8 url="https://example.com" title="Title">,
 @selector={:xpath=>"//uib-tab-heading[starts-with(normalize-space(text()), \"Accounts\") and contains(normalize-space(.), \"(2)\")]"}>
[3] pry> res.to_a
Selenium::WebDriver::Error::StaleElementReferenceError: Element not found in the cache - perhaps the page has changed since it was looked up
from [remote server] resource://fxdriver/modules/web-element-cache.js:9524:in `fxdriver.cache.getElementAt'
```

**Comment by titusfortner** [comment]
Has the number of elements in the collection decreased from initial evaluation?

**Comment by titusfortner** [comment]
I tried a few things, but can't replicate this. Please put this in your code before initializing the browser and paste the output:
`Selenium::WebDriver.logger.level = :info`

**Comment by maybelinot** [comment]
Selector is unique, thus number of elements is always 1

Browser initialization:
```
2017-06-30 12:59:18 INFO Selenium Skipping handshake as we know it is OSS.
2017-06-30 12:59:18 INFO Selenium -> POST session
2017-06-30 12:59:18 INFO Selenium    >>> http://127.0.0.1:7056/hub/session | {"desiredCapabilities":{"browserName":"firefox","version":"","platform":"ANY","javascriptEnabled":true,"cssSelectorsEnable
d":true,"takesScreenshot":true,"nativeEvents":false,"rotatable":false,"marionette":false},"capabilities":{"firstMatch":[{"browserName":"firefox"}]}}
2017-06-30 12:59:18 INFO Selenium <- {"name":"newSession","sessionId":"89a6ef87-8243-4d49-bd6e-07876b85c36f","status":0,"value":{"cssSelectorsEnabled":true,"browserName":"firefox","handlesAlerts":tru
e,"javascriptEnabled":true,"nativeEvents":false,"platform":"Linux","rotatable":false,"takesScreenshot":true,"version":"50.1.0","webStorageEnabled":true,"applicationCacheEnabled":true,"databaseEnabled
":true,"locationContextEnabled":true,"acceptSslCerts":true}}
2017-06-30 12:59:18 INFO Selenium Detected OSS dialect.
```
Failed part: 
```
2017-06-30 11:39:07 INFO Selenium -> POST session/33f053ba-2ef8-4078-96da-ac129c38d8a4/url
2017-06-30 11:39:07 INFO Selenium    >>> http://127.0.0.1:7056/hub/session/33f053ba-2ef8-4078-96da-ac129c38d8a4/url | {"url":"https://example.com"}
2017-06-30 11:27:18 INFO Selenium <- {"name":"get","sessionId":"e58c7c7f-03a7-4927-8e4d-2fed46a9ff09","status":0,"value":""}
2017-06-30 11:27:18 INFO Selenium -> POST session/e58c7c7f-03a7-4927-8e4d-2fed46a9ff09/elements
2017-06-30 11:27:18 INFO Selenium    >>> http://127.0.0.1:7056/hub/session/e58c7c7f-03a7-4927-8e4d-2fed46a9ff09/elements | {"using":"xpath","value":"//uib-tab-heading[starts-with(normalize-space(text
()), \"Service Accounts\") and contains(normalize-space(.), \"(2)\")]"}
2017-06-30 11:27:18 INFO Selenium <- {"name":"findElements","sessionId":"e58c7c7f-03a7-4927-8e4d-2fed46a9ff09","status":0,"value":[]}
2017-06-30 11:27:19 INFO Selenium -> POST session/e58c7c7f-03a7-4927-8e4d-2fed46a9ff09/elements
2017-06-30 11:27:19 INFO Selenium    >>> http://127.0.0.1:7056/hub/session/e58c7c7f-03a7-4927-8e4d-2fed46a9ff09/elements | {"using":"xpath","value":"//uib-tab-heading[starts-with(normalize-space(text
()), \"Service Accounts\") and contains(normalize-space(.), \"(2)\")]"}
2017-06-30 11:27:19 INFO Selenium <- {"name":"findElements","sessionId":"e58c7c7f-03a7-4927-8e4d-2fed46a9ff09","status":0,"value":[{"ELEMENT":"{8b4eaf16-f809-4993-a204-ab40becf7c6a}","element-6066-11
e4-a52e-4f735466cecf":"{8b4eaf16-f809-4993-a204-ab40becf7c6a}"}]}
2017-06-30 11:27:19 INFO Selenium -> GET session/e58c7c7f-03a7-4927-8e4d-2fed46a9ff09/element/%7B8b4eaf16-f809-4993-a204-ab40becf7c6a%7D/name
2017-06-30 11:27:20 INFO Selenium <- {"name":"getElementTagName","sessionId":"e58c7c7f-03a7-4927-8e4d-2fed46a9ff09","status":10,"value":{"message":"Element is no longer attached to the DOM","stackTra
ce":[{"methodName":"fxdriver.cache.getElementAt","fileName":"resource://fxdriver/modules/web-element-cache.js","lineNumber":9527,"columnNumber":24},{"methodName":"Utils.getElementAt","fileName":"file
:///tmp/webdriver-profile20170630-8436-dcus5q/extensions/fxdriver@googlecode.com/components/command-processor.js","lineNumber":9057,"columnNumber":10},{"methodName":"WebElement.getElementTagName","fi
leName":"file:///tmp/webdriver-profile20170630-8436-dcus5q/extensions/fxdriver@googlecode.com/components/command-processor.js","lineNumber":12340,"columnNumber":11},{"methodName":"DelayedCommand.prot
otype.executeInternal_/k","fileName":"file:///tmp/webdriver-profile20170630-8436-dcus5q/extensions/fxdriver@googlecode.com/components/command-processor.js","lineNumber":12877,"columnNumber":16},{"met
hodName":"DelayedCommand.prototype.executeInternal_","fileName":"file:///tmp/webdriver-profile20170630-8436-dcus5q/extensions/fxdriver@googlecode.com/components/command-processor.js","lineNumber":128
82,"columnNumber":7},{"methodName":"DelayedCommand.prototype.execute/<","fileName":"file:///tmp/webdriver-profile20170630-8436-dcus5q/extensions/fxdriver@googlecode.com/components/command-processor.j
s","lineNumber":12824,"columnNumber":5}]}}
```

**Comment by titusfortner** [comment]
`#to_a` caches (`@to_a ||= ...`), so it doesn't get re-evaluated when called. So the only way for `#to_a` to throw an exception is when it is first evaluated...

Oh that's odd... the `#find_elements` call is returning `{"ELEMENT":"{8b4eaf16-f809-4993-a204-ab40becf7c6a}"}` which shouldn't include the brackets inside the string. What version of Selenium is this? Was this working for you before? If so, what version Selenium & Watir?

**Comment by titusfortner** [comment]
Oh, this is Firefox. Update to the latest version of geckodriver & latest Firefox and see if that helps. Also is this going through a remote selenium server? If so what version?

**Comment by titusfortner** [comment]
@maybelinot do you have an update on this?

**Comment by maybelinot** [comment]
Hi @titusfortner, sorry for long silence
Unfortunately I wasn't able to check this on the latest(54) version of Firefox, getting [this](https://github.com/seleniumhq/selenium/issues/835) error when I'm trying to open a browser `Watir::Browser.new :firefox`. I'm running everything locally, no remote selenium servers. 
Also we using selenium-webdriver (3.4.3)

> Was this working for you before? If so, what version Selenium & Watir?

The problem appeared after we switched to 'watir' gem from 'watir-webdriver'

**Comment by titusfortner** [comment]
Looking deeper into this and some more info from Slack QA, I'm pretty sure this fixes the original issue: dfc9e112cd76ffdd495757d85141e36a76ee26ce. We'll get a release soon.

As for your last post, you should only get that if it was still trying to use the legacy firefox driver, which watir > 6 & selenium > 3.0 shouldn't do by default... Make sure you don't have any vestige of commonwatir, watir-classic, etc; otherwise I'm not sure.
